File: web/src/combinedIntelligence.ts
export type MlRiskLevel = 'Low' | 'Medium' | 'High'

export type CombinedIntelligenceResult = {
  finalRisk: MlRiskLevel
  finalAiInsight: string
}

/** Matches `pcod_risk_engine.ShaktiCarePCODRiskModel` rule tiers. */
export type MlFeatureRow = {
  cycle_length: number
  stress_level: number
  sleep_hours: number
  mood_score: number
  symptom_count: number
}

export type ShaktiAlertLevel = 'high' | 'medium' | 'normal'

export type ShaktiAlertItem = {
  level: ShaktiAlertLevel
  title: string
  detail?: string
}

export type IntelligenceThresholds = {
  /** Strictly above → High when fusing with Digital Twin. */
  highDeviation: number
  /** Above this and at most ``highDeviation`` → mild deviation (Medium path). */
  mildDeviation: number
}

const DEFAULT_THRESHOLDS: IntelligenceThresholds = {
  highDeviation: 5,
  mildDeviation: 2,
}

/** Rule-based ML risk (same thresholds as backend `ShaktiCarePCODRiskModel`). */
export function predictMlRiskFromFeatures(row: MlFeatureRow): MlRiskLevel {
  const highCycle = 35
  const highStress = 4
  const mediumCycle = 30
  const lowSleep = 5

  if (row.cycle_length > highCycle && row.stress_level > highStress) {
    return 'High'
  }
  if (row.cycle_length > mediumCycle || row.sleep_hours < lowSleep) {
    return 'Medium'
  }
  return 'Low'
}

/** Digital Twin baseline (see `DigitalTwinPage.tsx`). */
export const TWIN_BASELINE_CYCLE = 28
export const TWIN_BASELINE_MOOD = 4

/**
 * Twin deviation signals aligned with Digital Twin cards (cycle + stress + mood),
 * separate from the scalar score passed into `combineDigitalTwinAndMl`.
 */
export function computeTwinSignals(
  currentCycle: number,
  stress: number,
  mood: number,
): {
  cycleDeviationScore: number
  twinHigh: boolean
  twinMild: boolean
  twinNormal: boolean
  twinDeviation: boolean
} {
  const cycleDeviationScore = currentCycle - TWIN_BASELINE_CYCLE
  const moodDev = mood - TWIN_BASELINE_MOOD
  const twinHigh = cycleDeviationScore > 5 || stress >= 4
  const twinMild =
    !twinHigh && (cycleDeviationScore > 2 || Math.abs(moodDev) >= 1)
  const twinNormal = !twinHigh && !twinMild
  return {
    cycleDeviationScore,
    twinHigh,
    twinMild,
    twinNormal,
    twinDeviation: !twinNormal,
  }
}

function normalizeMlRisk(mlRiskLevel: string): MlRiskLevel {
  const key = mlRiskLevel.trim().toLowerCase()
  if (key === 'high') return 'High'
  if (key === 'medium') return 'Medium'
  if (key === 'low') return 'Low'
  throw new Error(`Invalid mlRiskLevel: ${JSON.stringify(mlRiskLevel)}`)
}

/**
 * ShaktiCare AI intelligence core: fuse Digital Twin deviation score with ML risk level.
 */
export function combineDigitalTwinAndMl(
  digitalTwinDeviationScore: number,
  mlRiskLevel: string,
  thresholds: Partial<IntelligenceThresholds> = {},
): CombinedIntelligenceResult {
  const t = { ...DEFAULT_THRESHOLDS, ...thresholds }
  const ml = normalizeMlRisk(mlRiskLevel)

  const highTwin = digitalTwinDeviationScore > t.highDeviation
  const mildTwin =
    digitalTwinDeviationScore > t.mildDeviation &&
    digitalTwinDeviationScore <= t.highDeviation

  let finalRisk: MlRiskLevel
  if (ml === 'High' || highTwin) {
    finalRisk = 'High'
  } else if (ml === 'Medium' || mildTwin) {
    finalRisk = 'Medium'
  } else {
    finalRisk = 'Low'
  }

  const finalAiInsight =
    finalRisk === 'Low'
      ? 'Your health patterns are stable and within healthy ranges'
      : 'Multiple factors are combining to suggest potential reproductive health considerations'

  return { finalRisk, finalAiInsight }
}

const MSG_COMBINED_HIGH = 'High stress and irregular cycles are contributing to elevated PCOD risk patterns'
const MSG_ML_EARLY = 'Behavioral factors like stress and sleep are signaling early reproductive health concerns'
const MSG_TWIN_LIFESTYLE = 'Lifestyle patterns are impacting hormonal balance and cycle regularity'

/**
 * Dashboard / ShaktiCare AI alerts: Digital Twin + ML together (copy + severity).
 */
export function buildShaktiCareCombinedAlerts(
  row: MlFeatureRow,
  thresholds: Partial<IntelligenceThresholds> = {},
): ShaktiAlertItem[] {
  const ml = predictMlRiskFromFeatures(row)
  const twin = computeTwinSignals(row.cycle_length, row.stress_level, row.mood_score)
  const fused = combineDigitalTwinAndMl(twin.cycleDeviationScore, ml, thresholds)

  const twinSummary = twin.twinDeviation
    ? twin.twinHigh
      ? 'Digital Twin: elevated deviation from baseline (cycle or stress).'
      : 'Digital Twin: mild shift from baseline (cycle or mood).'
    : 'Digital Twin: within normal baseline range.'

  const mlSummary =
    ml === 'High'
      ? 'ML: high behavioral / cycle risk band.'
      : ml === 'Medium'
        ? 'ML: moderate risk markers.'
        : 'ML: low risk band.'

  // 1) ML high while twin looks normal — behavioral early signal
  if (ml === 'High' && twin.twinNormal) {
    return [
      {
        level: 'high',
        title: MSG_ML_EARLY,
        detail: `${mlSummary} ${twinSummary}`,
      },
    ]
  }

  // 2) Twin deviation while ML is low — lifestyle / baseline shift
  if (ml === 'Low' && twin.twinDeviation) {
    const level: ShaktiAlertLevel =
      fused.finalRisk === 'High'
        ? 'high'
        : fused.finalRisk === 'Medium'
          ? 'medium'
          : 'medium'
    return [
      {
        level,
        title: MSG_TWIN_LIFESTYLE,
        detail: `${twinSummary} ${mlSummary} Combined assessment: ${fused.finalRisk}.`,
      },
    ]
  }

  // 3) General combined high (twin and/or ML per fusion)
  if (fused.finalRisk === 'High') {
    return [
      {
        level: 'high',
        title: MSG_COMBINED_HIGH,
        detail: `${twinSummary} ${mlSummary}`,
      },
    ]
  }

  if (fused.finalRisk === 'Medium') {
    return [
      {
        level: 'medium',
        title: 'Cycle irregularity combined with behavioral factors indicates moderate health trends',
        detail: `${twinSummary} ${mlSummary}`,
      },
    ]
  }

  return [
    {
      level: 'normal',
      title: 'Health patterns are within normal ranges',
      detail: `${twinSummary} ${mlSummary}`,
    },
  ]
}

File: combined_intelligence.py
from __future__ import annotations

from dataclasses import dataclass
from typing import TypedDict

from pcod_risk_engine import RiskLevel


class CombinedIntelligenceResult(TypedDict):
    """Core ShaktiCare AI output: fused Digital Twin + ML assessment."""

    final_risk: RiskLevel
    final_ai_insight: str


@dataclass(frozen=True)
class IntelligenceThresholds:
    """
    Digital Twin deviation score bands (same spirit as DigitalTwinPage.tsx).

    - Above ``high_deviation`` → strong twin signal (paired with ML in fusion rules).
    - Above ``mild_deviation`` up to and including ``high_deviation`` → mild deviation
      when ML alone would not already imply High.
    """

    high_deviation: float = 5.0
    mild_deviation: float = 2.0


def _normalize_ml_risk(ml_risk_level: str) -> RiskLevel:
    key = (ml_risk_level or "").strip().lower()
    if key == "high":
        return "High"
    if key == "medium":
        return "Medium"
    if key == "low":
        return "Low"
    raise ValueError(f"Invalid ml_risk_level: {ml_risk_level!r} (expected High, Medium, or Low)")


def combine_digital_twin_and_ml(
    *,
    digital_twin_deviation_score: float,
    ml_risk_level: str,
    thresholds: IntelligenceThresholds | None = None,
) -> CombinedIntelligenceResult:
    """
    ShaktiCare AI intelligence core: merge Digital Twin deviation with ML risk.

    Logic:
    - Final High if ML is High OR twin deviation is strictly above the high threshold.
    - Else Final Medium if ML is Medium OR mild deviation (between mild and high band).
    - Else Final Low.

    Narrative:
    - Non-Low → twin + ML risk alignment copy.
    - Low → normal behavior across both systems.
    """
    t = thresholds or IntelligenceThresholds()
    ml = _normalize_ml_risk(ml_risk_level)

    high_twin = digital_twin_deviation_score > t.high_deviation
    mild_twin = (
        digital_twin_deviation_score > t.mild_deviation
        and digital_twin_deviation_score <= t.high_deviation
    )

    if ml == "High" or high_twin:
        final_risk: RiskLevel = "High"
    elif ml == "Medium" or mild_twin:
        final_risk = "Medium"
    else:
        final_risk = "Low"

    if final_risk == "Low":
        insight = "Normal behavior detected across both systems"
    else:
        insight = "Digital Twin shows deviation + ML model confirms risk pattern"

    return {"final_risk": final_risk, "final_ai_insight": insight}
